ce/api/streamflow/watershed.py

```diff
@@ -0,0 +1,421 @@
+              if neighbour not in visited and is_upstream(neighbour, cell))
+        )
+
+    if debug:
+        return upstream(target), {'time': time.time() - start_time}
+    return upstream(target)
+
+
+def VIC_direction_matrix(lat_step, lon_step):
+    """ Return a VIC direction matrix, which is a matrix indexed by the VIC
+    streamflow direction codes 0...9, with the value at index `i` indicating
+    the offsets from the data index in a streamflow file required to
+    step in that streamflow direction.
+
+    :param lat_step: Difference between two successive latitudes in streamflow
+        file. (Only) sign matters.
+    :param lon_step: Difference between two successive longitudes in streamflow
+        file. (Only) sign matters.
+    :return: tuple of offset pairs
+
+    The offsets must account for the sign of the step in the lat and lon
+    dimensions in the streamflow file.
+    For example, in a streamflow file with lat and lon both increasing with
+    increasing index, the offset to step northeast is [1, 1].
+
+    Note that argument order is (lat, lon), not (lon, lat).
+    """
+    base = (
+        ( 0,  0),   # filler - 0 is not used in the encoding
+        ( 1,  0),   # 1 = north
+        ( 1,  1),   # 2 = northeast
+        ( 0,  1),   # 3 = east
+        (-1,  1),   # 4 = southeast
+        (-1,  0),   # 5 = south
+        (-1, -1),   # 6 = southwest
+        ( 0, -1),   # 7 = west
+        ( 1, -1),   # 8 = northwest
+        ( 0,  0),   # 9 = outlet
+    )
+    lat_dir = int(math.copysign(1, lat_step))
+    lon_dir = int(math.copysign(1, lon_step))
+    return tuple(
+        (lat_dir * lat_base, lon_dir * lon_base)
+        for lat_base, lon_base in base
+    )
+
+
+def get_time_invariant_variable_dataset(sesh, ensemble_name, variable):
+    """Locates and opens a time-invariant dataset.
+    These datasets contain things like elevation or area of a grid cell -
+    they're independent of time and there should be only one per ensemble.
+    If more or less than one is found in the ensemble, it raises an error
+
+    :param sesh: (sqlalchemy.orm.session.Session) A database Session object
+    :param ensemble_name: Name of the ensemble containing data files
+    :param variable: Name of *variable* inside dataset. This is not the
+        dataset filename or unique id.
+    :return: (netcdf.Dataset) netcdf Dataset object representing the file.
+    """
+    query = (
+        sesh.query(distinct(DataFile.filename).label('filename'))
+        .join(
+            DataFileVariable,
+            EnsembleDataFileVariables,
+            Ensemble,
+        )
+        .filter(Ensemble.name == ensemble_name)
+        .filter(DataFileVariable.netcdf_variable_name == variable)
+        .filter(DataFile.time_set_id == None)
+    )
+
+    file = query.one()  # Raises exception if n != 1 results found
+    return Dataset(file.filename, 'r')
+
+
+def hypsometry(elevations, areas, bin_start=0, bin_width=100, num_bins=46):
```

Comment:
I don't think this indicates a problem in either the data or the code. 

In the elevation file we have (presumably from HI), we see 

```
$ ncdump -h elevation.nc 
netcdf elevation {
dimensions:
	lon = 1088 ;
	lat = 512 ;
variables:
	double lon(lon) ;
		lon:standard_name = "longitude" ;
		lon:long_name = "longitude" ;
		lon:units = "degrees_east" ;
		lon:axis = "X" ;
	double lat(lat) ;
		lat:standard_name = "latitude" ;
		lat:long_name = "latitude" ;
		lat:units = "degrees_north" ;
		lat:axis = "Y" ;
	float elev(lat, lon) ;
		elev:long_name = "Average elevation of grid cell" ;
		elev:units = "m" ;
		elev:_FillValue = NaNf ;
		elev:missing_value = NaNf ;
		elev:cell_methods = "area: mean" ;
```

I'm taking `elev:long_name = "Average elevation of grid cell" ` at face value. In which case, skipping elevation bands is not contradictory: Consider a case where the area in question contains lowlands and highlands with very steep changes between them. Then the set of average elevations could easily jump from, say, 100 m to 500 m, missing the intervening elevation bands. This will happen wherever the topography varies wildly. We have a lot of that in BC.

Reply:
This is much more likely to occur when there are just a few cells in the watershed, since `n` cells can represent at most `n` elevations, and the possibility that there is a steep area whose missing elevation bands are not filled in by some other area is much greater. The example you give has 9 cells, so I'd conclude that this is exactly what we might expect.